[MRG] stc plotting tutorial

jasmainak opened the pull request:
This is just to declare my intent ...

I'll finish the PR tomorrow

jona-sassenhagen commented:
Good to have this.

jasmainak commented:
Okay, so I think I have a first version ready here: https://9480-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_visualize_stc.html

jasmainak commented:
Actually I think this is good to go from my end. Unless I missed something ...

agramfort commented on tutorials/plot_visualize_stc.py:
```diff
@@ -0,0 +1,137 @@
+###############################################################################
+# This too comes with a convenient plot method.
+
+stc.plot(src, subject='sample', subjects_dir=subjects_dir)
```
please also illustrate the matplotlib backend

agramfort commented on tutorials/plot_visualize_stc.py:
```diff
@@ -0,0 +1,137 @@
+# objects. Note that for this visualization to work, you must have ``mayavi``
+# and ``pysurfer`` installed on your machine.
+initial_time = 0.1
+stc.plot(subjects_dir=subjects_dir, initial_time=initial_time)
```
please also illustrate the matplotlib backend

agramfort commented on tutorials/plot_visualize_stc.py:
```diff
@@ -0,0 +1,137 @@
+
+##############################################################################
+# Finally, we can visualize the dipole.
+dip.plot_locations(fname_trans, 'sample', subjects_dir, mode='orthoview')
```
anything else than orthoview to showcase?

jasmainak commented:
@agramfort I updated the tutorial to add `matplotlib` viz and also removed the argument `"orthoview"` since that's the only one that's supported. However, the `mayavi` and the `matplotlib` figure look quite different to me.

![mayavi](https://user-images.githubusercontent.com/15852194/46091599-2d509d80-c181-11e8-8bb5-19e6d07e106b.png)
![matplotlib](https://user-images.githubusercontent.com/15852194/46091614-33467e80-c181-11e8-9e2f-7a190b06451e.png)

The colorbar is missing + the brain looks flipped. Is this okay for you?

agramfort commented:
Not ok ... can you look into it ?

jasmainak commented:
@agramfort I fixed the issue of different views. See https://github.com/mne-tools/mne-python/pull/5528/commits/9cf68682080aaff4a4f85e26a5fa873b4a8218f4 The colormap is still not equivalent. To fix that, I think we should first add a colorbar (which is not currently supported with mpl backend) and then play around with the colormap. Do you want me to do this in this PR itself?

jasmainak commented:
see https://9509-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_visualize_stc.html

jasmainak commented:
Build failures are unrelated to PR

agramfort commented:
@jasmainak please yes do it in this PR as we can directly see how nice it is. thanks heaps

jasmainak commented:
Alright. Getting closer. [Colorbar added](https://9522-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_visualize_stc.html). Not sure why the maps are still not the same though ...

larsoner commented:
Those maps look the same to me, aside from expected blending and decimation differences. The colorbar rendering in Mayavi is broken but that's not new to this PR.

jasmainak commented:
@larsoner if you say that the colorbar rendering of mayavi is not correct, then we are fine. This is what I see on my local machine:

![comparison](https://user-images.githubusercontent.com/15852194/46188949-0c3f9800-c2ba-11e8-9e7e-1a81287b6422.png)

Ignore the fact that the brain maps don't look exactly the same -- it's related to the transparency issue on mayavi that I have. However, the colorbar still seems a bit different, don't you think? The mayavi colorbar seems to squeeze the red part of the "hot" colormap more than the matplotlib colorbar.

I tried a couple of things but no success. Using the pysurfer private function, [_scale_sequential_lut](https://github.com/nipy/PySurfer/blob/master/surfer/viz.py#L2768), I was able to get a similar colorbar but then the brainmap looks different (and full of yellow). So, either you are correct that `mayavi` figure is broken. Or there is still something to fix ...

agramfort commented:
nice ! can you fix the CIs issues?

larsoner commented:
Ahh I see what you mean.

You should be able to use the new unified function, it should already do the `cmap` "squeezing". My guess is that in the `mpl` code it is being squeezed again, which is not what we want.

larsoner commented on mne/viz/_3d.py:
```diff
@@ -1552,6 +1568,20 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
+        cb_yticks = plt.getp(cax, 'yticklabels')
+        plt.setp(cb_yticks, color='w')
+        cax.tick_params(labelsize=16)
+        cb.patch.set_facecolor((0., 0., 0., 1.0))
```
Use middle gray (`'0.5'`) not black

jasmainak replied:
It seems Pysurfer uses black. Don't we want to replicate that?

larsoner replied:
It doesn't on master, it uses mid gray (or at least is coded to do so)

larsoner commented on mne/viz/_3d.py:
```diff
@@ -1460,31 +1460,47 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
     ctrl_pts, colormap, _, _ = _limits_to_control_points(
-        clim, stc.data, colormap, transparent=False, fmt='matplotlib')
+        clim, stc.data, colormap, transparent=transparent, fmt='matplotlib')
+    colormap = plt.get_cmap(colormap)
```
You shouldn't need to do this anymore, the output should already be a `cmap` after my recent changes

larsoner commented on mne/viz/_3d.py:
```diff
@@ -1460,31 +1460,47 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
+
+    locs = np.clip(np.linspace(ctrl_pts[0] / ctrl_pts[2], 1, 256), 0, 1)
+    colormap = colormap(locs)
+    colormap = colors.ListedColormap(colormap)
```
You also should not need to do this, the `_limits_to_control_points` should already do the correct nonlinear scaling

jasmainak replied:
I tried the following:

```py

    ctrl_norm = ctrl_pts - ctrl_pts[0]
    ctrl_norm /= ctrl_norm[2]
    vals = np.interp(np.linspace(0, 1, 256), ctrl_norm, [0., 0.5, 1.])
    colormap = np.array(colormap(vals))
    if transparent:
        colormap[:, 3] = np.interp(np.linspace(0, 1, 256), ctrl_norm,
                                   [0, 1, 1])
    colormap = colors.ListedColormap(colormap)
```

but for some strange reason, the brainmap starts saturating then:

![brainmap](https://user-images.githubusercontent.com/15852194/46231932-9a0f9780-c33b-11e8-9582-474ca40f4e55.png)

I can't seem to get both the colorbar and brainmap to look good ...

jasmainak commented on mne/viz/_3d.py:
```diff
@@ -1552,6 +1571,21 @@ def _plot_mpl_stc(stc, subject=None, surface='inflated', hemi='lh',
+    # add colorbar
+    from mpl_toolkits.axes_grid1.inset_locator import inset_axes
+    sm = plt.cm.ScalarMappable(cmap=cmap,
+                               norm=plt.Normalize(0, ctrl_pts[2]))
```
okay I _think_ I got it to work. But there is a hack here that I don't understand at the moment. There is probably a better way to do this ...

agramfort commented:
@jasmainak please fix the CIs thanks

jasmainak commented on mne/viz/_3d.py:
```diff
@@ -1441,12 +1440,12 @@ def _smooth_plot(this_time, params):
+    colors[:, :3] *= colors[:, [3]]
+    colors[:, :3] += greymap(curv_ave)[:, :3] * (1.  - colors[:, [3]])
+    colors[:, 3] = 1.
+    facecolors[:] = colors
```
@larsoner can you explain what you fixed?

larsoner replied:
The alpha blending was wrong before

larsoner replied:
(This is now equivalent to standard OpenGL blending like ONE_MINUS_SRC_ALPHA)

jasmainak commented:
@larsoner are you done? The pysurfer and matplotlib maps still look somewhat different though ...

![pysurfer](https://user-images.githubusercontent.com/15852194/46310000-1ef0f000-c58c-11e8-9c54-6a644d25cf18.png)

![matplotlib](https://user-images.githubusercontent.com/15852194/46310017-27492b00-c58c-11e8-8490-a3e025754ef6.png)

jasmainak commented:
oh wait, and now there is no colorbar in matplotlib ...

larsoner commented:
No I am not done

larsoner commented:
Should work now. Locally I see this:

![screenshot from 2018-10-01 16-21-04](https://user-images.githubusercontent.com/2365790/46313519-0ede0e00-c596-11e8-85a0-65bc961b343f.png)

![screenshot from 2018-10-01 16-21-06](https://user-images.githubusercontent.com/2365790/46313525-11406800-c596-11e8-9112-7df13f03ad80.png)

larsoner commented:
Updated examples:

https://9547-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_visualize_stc.html

https://9547-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_lcmv_beamformer_volume.html

https://9547-1301584-gh.circle-artifacts.com/0/html/auto_tutorials/plot_mne_dspm_source_localization.html

jasmainak commented:
@larsoner the CIs are failing and for the same reason the tutorial is failing on my computer. I don't have `master` of pysurfer ...

jasmainak commented:
But the visualizations look great now !

agramfort commented:
looks great but we have now these mayavi colorbar issues.

do you see them too?

larsoner commented:
Not locally. They are on every PR and in `master` and probably a pain to fix (CI reconfiguration / MESA bug?) so probably shouldn't be handled here.

agramfort commented:
ok then. Besides the CIs issues LGTM.

thx heaps @larsoner

larsoner commented:
I'll fix the CIs in a few hours

agramfort commented:
merge when green.

thx @jasmainak @larsoner
